tests/func/experiments/test_remote.py

```diff
@@ -121,12 +121,13 @@ def test_list_remote(tmp_dir, scm, dvc, git_downstream, exp_stage, use_url):
     downstream_exp = git_downstream.tmp_dir.dvc.experiments
     assert downstream_exp.ls(git_remote=remote) == {}
 
-    exp_list = downstream_exp.ls(rev=baseline_a, git_remote=remote)
+    git_downstream.tmp_dir.scm.fetch_refspecs(remote, ["master:master"])
```

Comment:
One question here, currently the rev collection is on the local side. And the git tree in local might not equal to the remote. It is better to collect the revs from the remote directly, but we don't have a related `scm` function yet?

Reply:
I think for now we should not support using `--all-branches/--all-tags` with remote exp listing. Only supporting the existing use case (with `--rev` and `--all-commits` in place of `--all`) for remote listing should be fine for now.

Reply:
My opinion:
1. `pull` and `push` are two sides of one coin, should stay the same here. 
2. `list` can be used for previews on other commands that should support all of the flags here.
3.  for `show`, `num` can be used to show the results improving progress,   `all_branches`, `all_tags`, `all_commits` are also useful in showing the different versions of models.
4. `pull` and `push` should support all of the flags in `show` for the experiments sharing.
5. `rev` is useful for `remove` to clean up the result after we fixed the parameters. `all_commits` which is supported now is used to clean up a repo. But I don't think that `all_tags` and `all_branches` have any actual use case here, same with `num` .

A summary:

|  cmd  |  list  |  show  |  pull/push  |  remove  |
|---|---|---|---|---|
|  --rev | yes  | yes  |  yes  |  yes  |
|  --num | yes  | yes  |  yes  |  no  |
|  --all_commits |  yes  | yes |  yes  | yes  |
|  --all_tags  |  yes  |  yes |  yes  |  no  |
|  --all_branches  |  yes  | yes  |  yes  |  no  |

@dberenbaum  What do you think of it?

Reply:
😅 Thanks for the great synopsis @karajan1001!

Seems like the only questionable part is the flags for pull/push, mainly because some of them seem somewhat unlikely and potentially very long-running. I don't have a strong opinion on `--num`, `--all-commits`, `--all-tags`, `--all-branches` for push/pull. It's possible but unlikely that someone would do `dvc exp pull --all-tags; dvc exp show --all-tags`. Whatever you and @pmrowla decide on those is fine with me.

Reply:
The issue is that `--all-tags`/`--all-branches`/`--all-commits` is ambiguous in the remote command use case. It is unclear whether the command should use the local repo's refs or the remote repo's refs in order to determine what constitutes "all tags/branches/commits", because they will not always be the same thing.

I may have a lot of local dev branches that I don't want pushed to a github repo, and the github repo may have things like release/backport/testing branches that I don't want to pull into my local copy.

So when I do `dvc exp push --all-branches`, does that mean I want to push all experiments based on all of my local branches (including ones that I have not also pushed to github)?

And on `dvc exp pull --all-branches`, does that mean I want to pull experiments from branches which only exist in the remote (and that I have not also `git pulled` locally)?

Reply:
IMO, push/pull should be explicit, and in the branch/tag case we should require the user to use `--rev branch-name`. If they want to push/pull for all commits in that branch they can use `--rev branch-name -n -1`

Reply:
Agreed, `tags` and `branches` might not be synchronized between remote and local, we should not support them in the remote use case, then no need in `list` also. I'll implement the following version.

|  cmd  |  list  |  show  |  pull/push  |  remove  |
|---|---|---|---|---|
|  --rev | yes  | yes  |  yes  |  yes  |
|  --num | yes  | yes  |  yes  |  yes  |
|  --all_commits |  yes  | yes |  yes  | yes  |
|  --all_tags  |  no  |  yes |  no  |  no  |
|  --all_branches  |  no  | yes  |  no  |  no  |

Reply:
So what happens if `push/pull --all-commits` includes branches that aren't on one of local/remote? It's useful for `list` since that only operates on the remote. `push/pull` operate on both, so maybe we should remove `push/pull --all-commits`?

Also, no strong opinion, but any reason why you decided to add support for `remove --num` in the updated table?

Reply:
> So what happens if push/pull --all-commits includes branches that aren't on one of local/remote? It's useful for list since that only operates on the remote. push/pull operate on both, so maybe we should remove push/pull --all-commits?

The original solution follows the from repo (local in `push`, remote in `pull`), the same with the results in `list`. But for now, an experiment whose baseline doesn't exist in the repo will be limited on many commands. 

> Also, no strong opinion, but any reason why you decided to add support for remove --num in the updated table?

No very strong reasons, only because:
 1. It might be useful when cleaning a branch of experiments, and it is harmless.
 2. Forbiddening requires extra code, which is unnecessary.